src/http/matcher/mod.rs

```diff
@@ -880,6 +696,309 @@ impl<State, Body> crate::service::Matcher<State, Request<Body>> for HttpMatcherK
+            ("GET", Version::HTTP_2, "www.lorem.com", false),
+            ("GET", Version::HTTP_3, "www.example.com", false),
+            ("GET", Version::HTTP_3, "www.lorem.com", true),
+        ];
+
+        for (method, version, uri, expected) in cases {
+            let req = Request::builder()
+                .method(method)
+                .version(version)
+                .uri(uri)
+                .body(())
+                .unwrap();
+
+            assert_eq!(
+                matcher.matches(None, &Context::default(), &req),
+                expected,
+                "({:#?}).matches({:#?})",
+                matcher,
+                req
+            );
+        }
+    }
+
+    #[test]
+    fn test_matcher_or_and_or_and_negation() {
+        let matcher = (HttpMatcher::method_post().or(HttpMatcher::method_get()))
+            .and(
+                HttpMatcher::version(VersionMatcher::HTTP_2)
+                    .or(HttpMatcher::version(VersionMatcher::HTTP_3)),
+            )
+            .and(HttpMatcher::domain("www.example.com").negate());
+
+        let cases = [
+            ("POST", Version::HTTP_2, "www.lorem.com", true),
+            ("POST", Version::HTTP_2, "www.example.com", false),
+            ("POST", Version::HTTP_3, "www.lorem.com", true),
+            ("POST", Version::HTTP_3, "www.example.com", false),
+            ("POST", Version::HTTP_09, "www.lorem.com", false),
+            ("POST", Version::HTTP_09, "www.example.com", false),
+            ("GET", Version::HTTP_2, "www.lorem.com", true),
+            ("GET", Version::HTTP_2, "www.example.com", false),
+            ("GET", Version::HTTP_3, "www.lorem.com", true),
+            ("GET", Version::HTTP_3, "www.example.com", false),
+            ("GET", Version::HTTP_09, "www.lorem.com", false),
+            ("GET", Version::HTTP_09, "www.example.com", false),
+            ("PUT", Version::HTTP_2, "www.lorem.com", false),
+            ("PUT", Version::HTTP_2, "www.example.com", false),
+            ("PUT", Version::HTTP_3, "www.lorem.com", false),
+            ("PUT", Version::HTTP_3, "www.example.com", false),
+            ("PUT", Version::HTTP_09, "www.lorem.com", false),
+            ("PUT", Version::HTTP_09, "www.example.com", false),
+        ];
+
+        for (method, version, uri, expected) in cases {
+            let req = Request::builder()
+                .method(method)
+                .version(version)
+                .uri(uri)
+                .body(())
+                .unwrap();
+
+            assert_eq!(
+                matcher.matches(None, &Context::default(), &req),
+                expected,
+                "({:#?}).matches({:#?})",
+                matcher,
+                req
+            );
+        }
+    }
+
+    #[test]
+    fn test_matcher_negation_and_header_contains() {
+        let matcher = HttpMatcher::method_get().negate().and_header_contains(
+            "content-type".parse().unwrap(),
```

Comment:
You can use `HeaderName/HeaderValue`::from_static for these cases, just make sure to keep the names lower case as otherwise you get a panic, but that's okay for your cases.